render/naga-agal/src/builder.rs

```diff
@@ -808,43 +808,115 @@ impl<'a> NagaBuilder<'a> {
         extend_to_vec4: bool,
         output: VectorSize,
     ) -> Result<Handle<Expression>> {
-        let (mut base_expr, source_type) = match source.register_type {
-            // We can use a function argument directly - we don't need
-            // a separate Expression::Load
-            RegisterType::Attribute => (
-                self.get_vertex_input(source.reg_num as usize)?,
-                self.shader_config.vertex_attributes[source.reg_num as usize]
-                    .ok_or(Error::MissingVertexAttributeData(source.reg_num as usize))?,
-            ),
-            RegisterType::Varying => (
-                self.emit_varying_load(source.reg_num as usize)?,
-                VertexAttributeFormat::Float4,
-            ),
-            RegisterType::Constant => (
-                self.emit_const_register_load(source.reg_num as usize)?,
-                // Constants are always a vec4<f32>
-                VertexAttributeFormat::Float4,
-            ),
-            RegisterType::Temporary => {
-                let temp = self.get_temporary_register(source.reg_num as usize)?;
-                (
-                    self.evaluate_expr(Expression::Load { pointer: temp }),
+        let mut load_register = |register_type: &RegisterType, reg_num| {
+            match register_type {
```

Comment:
(Very minor nit: Could move `reg_num` into its own `let reg_num = reg_num as usize;` before the match-branches, though this will likely be cleaner to do in a future PR).

Reply:
I've changed the closure to take `reg_num` as a `usize`.